A few changes to the original code

AnakinFoxe opened the pull request:
First of all, thanks for this great script. It helped me download all the albums for my kid 😆 

Made a few changes to the original code:
- Authenticate with password (with `-p PASSWORD` option), however if session ID is provided, password will be ignored.
- Add an option `--folder` to download all the albums under the specific folder
- Concurrently download images of the same album
- A few useful refactors (I think)
  - Split codes into functions and organize under class `SmugMugDownloader`
  - Use one session for all the requests
  - Use `logging`
  - Use `@retry` to manage retry for `get_json()`
  - ~~Use `Makefile` to setup a virtual environment to work with~~

AnakinFoxe committed:
Authenticate with password

AnakinFoxe committed:
Merge pull request #1 from AnakinFoxe/authenticate-with-password

Authenticate with password

AnakinFoxe committed:
Download all the albums under specific folder

AnakinFoxe committed:
Refactor the code

AnakinFoxe committed:
Concurrently download images in an album

AnakinFoxe committed:
Update readme to reflect the changes

AnakinFoxe committed:
Small update to readme

AnakinFoxe committed:
Add ensurepip

andyjsmith commented:
Sorry for the late reply, I appreciate the changes!

I'm a bit hesitant on the password authentication due to some unhandled edge cases (e.g. what if the user has 2FA enabled?). Maybe we could keep both the user/pass and session token auth mechanisms instead of replacing it. As for the Makefile, I would like to keep this script platform-independent so I don't want to merge in a Unix-specific script. At some point a setup.py / pyproject.toml may be appropriate.

The other contributions are great but they probably need to be split off as separate PRs.

jason-fer reviewed (commented):
This repo is a lifesaver. Good point on allowing 2 separate modes Andy.

AnakinFoxe committed:
Merge remote-tracking branch 'upstream/master'

AnakinFoxe committed:
Keep the -s SESSION argument

AnakinFoxe committed:
Remove Makefile

AnakinFoxe commented:
@andyjsmith thanks for the comments! I have updated the PR with following changes (updated the PR description as well)
- Kept the `-s SESSION` argument. If session ID is provided, password will be ignored.
- Removed Makefile
- README.md has been updated accordingly.
